sway/input/text_input.c

```diff
@@ -606,16 +620,28 @@ void sway_input_method_relay_init(struct sway_seat *seat,
 	relay->text_input_new.notify = relay_handle_text_input;
 	wl_signal_add(&server.text_input->events.text_input,
 		&relay->text_input_new);
+	relay->text_input_manager_destroy.notify = relay_handle_text_input_manager_destroy;
+	wl_signal_add(&server.text_input->events.destroy,
+		&relay->text_input_manager_destroy);
 
 	relay->input_method_new.notify = relay_handle_input_method;
 	wl_signal_add(
 		&server.input_method->events.input_method,
 		&relay->input_method_new);
+	relay->input_method_manager_destroy.notify = relay_handle_input_method_manager_destroy;
+	wl_signal_add(&server.input_method->events.destroy,
+		&relay->input_method_manager_destroy);
 }
 
 void sway_input_method_relay_finish(struct sway_input_method_relay *relay) {
+	// return early if finish was already called
+	// can be called due to seat or manager protocol object being destroyed
+	if (!relay->input_method_new.link.prev) return;
```

Comment:
Hm, this sounds like a workaround.

Since `server.text_input` and `server.input_method` are only accessed in `sway_input_method_relay_init()`, we could instead do something like:

```c

static void relay_handle_text_input_manager_destroy(struct wl_listener *listener, void *data) {
	struct sway_input_method_relay *relay = wl_container_of(listener, relay,
		text_input_manager_destroy);
	wl_list_remove(&relay->input_method_new.link);
	wl_list_remove(&relay->input_method_manager_destroy.link);
	wl_list_init(&relay->input_method_new.link);
	wl_list_init(&relay->input_method_manager_destroy.link);
}
```

Perhaps this can be extracted to a helper function `sway_input_method_relay_finish_input_method()` and re-used in `sway_input_method_relay_finish()`.

Reply:
I wrote this before I knew `wl_list_init()` could be used to make this work.

Thanks, I'll look into this.

Reply:
Implemented this exactly as suggested, using a new helper `sway_input_method_relay_finish_input_method()` that is called by `relay_handle_text_input_manager_destroy()` and `sway_input_method_relay_finish()`.

Done in 777ecd9e.

Reply:
Shouldn't `relay_handle_text_input_manager_destroy()` remove the listeners for `text_input_new` and `text_input_manager_destroy`, rather than the input method listeners?

Reply:
Yes, my bad, my suggestion has a typo!